spec/core/ics-004-channel-and-packet-semantics/LOCALHOST.md

```diff
@@ -0,0 +1,72 @@
+# Localhost Support in 04-channel
+
+## Rationale
+
+IBC provide the ICS-26 interface for two mutually-untrusted applications to negotiate a communication channel and start sending messages to each other. This interface is now widely adopted by applications (both smart contracts and static modules) so that they can speak to applications on remote chains. However, there is no way to reuse this interface to communicate with applications on the same chain. So applications that wish to communicate with counterparties on remote and local state machines must for the moment have two separate implementations to handle both cases. This is a dev-UX issue that we can solve by providing localhost functionality (analogous to localhost for the internet protocol).
+
+One approach is to implement this feature at the client layer. However this leads to a number of inefficiencies. Firstly, it requires the creation of a connection (handshake messages mediated by relayers) to be sent from the state machine back to itself in order to connect. It then requires that state that exists at the channel layer be routed all the way into the client so that it can then be inspected for verification purposes. Lastly, it does not immediately lend itself to future improvements that might make localhost packet flows atomic.
+
+The state that applications wish to verify about their counterparties are all stored in the channel store (i.e. `channelEnd`, `packet_commitment`, `receipt`, `acknowledgement`). Thus, the simplest approach is to add the special localhost handling here at the channel layer, rather than propogating it all the way down the stack. This provides the uniform interface that applications are looking for with minimal changes to the core handlers.
+
+## Technical Specification
+
+Implementations will reserve a special connectionID: `connection-localhost` which applications can build their channels on in order to communicate with a counterparty on the same state machine.
+
+In all channel verification methods, the channel logic must first check that the connectionID is the sentinel localhost connectionID: `connection-localhost`. If it is, then we simply introspect our own channel store to check if the expected state is stored at the expected key path. If the connection is not a localhost, then we delegate verification to the connection layer and ultimately the counterparty client so it can verify that the expected state is stored at the expected key path on the remote chain.
```

Comment:
I like this idea. It assumes we have different (port, channel, connection) ids on both sides, right?

What if someone makes a channel between `transfer` and `transfer` on `connection-localhost`?
Both ends would have the same port and channel ID and it would be impossible to distinguish the two sides of the channel (when inspecting own store).

I think we need something to handle this case (even if that is making such channels illegal and refusing to create them)

Reply:
> I like this idea. It assumes we have different (port, channel, connection) ids on both sides, right?

The connection ids would be `connection-localhost` on both sides so really just assuming different channel ids. 

> What if someone makes a channel between `transfer` and `transfer` on `connection-localhost`? Both ends would have the same port and channel ID and it would be impossible to distinguish the two sides of the channel (when inspecting own store).

The connections don't actually exist in state, you just check if the connection id is set to `connection-localhost` and since both ends are on the same chain the calls to `GenerateChannelIdentifier` in `ChanOpenInit` & `ChanOpenTry` should ensure that we end up with different channel ids.

Reply:
@jtieri is correct

Reply:
> What if someone makes a channel between transfer and transfer on connection-localhost?

They would have different channelIDs, but the same portID and connectionID

Reply:
Could you add this as a testcase @jtieri if it doesn't already exist just to prove the above assertion?

Reply:
If they get unique channel-ids, perfect.

I was afraid this would be one channel, but yeah, if they are constructed separately as if they are different chains, they should be eg. channel-7 and channel-8

A test would be good though, as this is a lot of reasoning about a key correctness issue.

Reply:
btw, I really like this idea. Much simpler and finally adding localhost will be nice